skcosmo/preprocessing/KernelFlexibleCenterer.py

```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import sklearn as sk
+from sklearn.base import TransformerMixin, BaseEstimator
+
+
+class KernelFlexibleCenterer(TransformerMixin, BaseEstimator):
+    """Kernel centering method, similar to KernelCenterer,
+    but with additional parameters, relative to which centering
+    is carried out:
+    ref_cmean: an array with means for each column.
+    ref_rmean: an array with means for each row.
+    ref_mean: an average for the whole kernel matrix
+    reference: a reference, relative to which centering is carried out"""
+
+    def __init__(
+        self, *, reference=None, ref_cmean=None, ref_rmean=None, ref_mean=None
+    ):
+        """Initialize KernelFlexibleCenterer. Specifies the parameters relative to which the  kernel
+        matrix will be centered."""
+
+        self.reference = reference
+        self.ref_cmean = ref_cmean
+        self.ref_rmean = ref_rmean
+        self.ref_mean = ref_mean
+
+    def fit(self, K, y=None):
+        """Fit KernelFlexibleCenterer
+        Parameters
+        ----------
+        K : ndarray of shape (n_samples, n_samples)
+            Kernel matrix.
+        y : None
+            Ignored.
+        Returns
+        -------
+        self : object
+            Fitted transformer.
+        """
+        K_ref = self.reference
+        if K_ref is None:
+            K_ref = K
+            self.reference = K
+        else:
+            if K.shape[1] != K_ref.shape[0]:
+                raise ValueError(
+                    "The kernel to be centered and the reference have inconsistent sizes"
+                )
+        if K_ref.shape[0] != K_ref.shape[1]:
+            raise ValueError(
+                "The reference kernel is not square, and does not define a RKHS"
+            )
+
+        if self.ref_cmean is None:
+            self.ref_cmean = K_ref.mean(axis=0)
+
+        if self.ref_rmean is None:
+            self.ref_rmean = K_ref.mean(axis=1)
+
+        if self.ref_mean is None:
+            self.ref_mean = K_ref.mean()
+        return self
```

Comment:
I agree with the first version of the code but disagree with the second. For example, we instantiated a class with only self.reference and after that, we do .fit (). In the first code, our .fit (), using self.reference, will complete the self.ref_mean, self.ref_rmean and self.ref_cmean instance. According to the second code, after .fit (), we generally "forget" about what we were doing during the creation of the class instance.

Reply:
For self.return exist the reason. We should have the possibility to write smth like that:
transformer = KernelFlexibleCenterer().fit(K)
Without self.return we will have None.

Reply:
> I agree with the first version of the code but disagree with the second. For example, we instantiated a class with only self.reference and after that, we do .fit (). In the first code, our .fit (), using self.reference, will complete the self.ref_mean, self.ref_rmean and self.ref_cmean instance. According to the second code, after .fit (), we generally "forget" about what we were doing during the creation of the class instance.
> Maybe it makes sense then to give the opportunity to set the centering parameters during .fit ()? Passing them as parameters.

In general we should reserve "fitting" the transformer (i.e. setting reference, ref_cmean, etc.) for the `fit` function alone. I think we should take those parameters out of the constructor and then make `fit` look like this:

```
def fit(self, K=none, ref_cmean=None, ref_rmean=None, ref_mean=None):

    if K is not None:
        if K.shape[0] != K.shape[1]:
                    raise ValueError(
                        "The reference kernel is not square, and does not define a RKHS"
                    )
        self.reference_shape = K.shape # because we only ever need its shape later on
        self.ref_cmean = K.mean(axis=0)
        self.ref_rmean = K.mean(axis=1)
        self.ref_mean = K.mean()
    else:
        assert ref_cmean is not None and ref_rmean is not None and ref_mean is not None

        if len(ref_rmean) != len(ref_cmean):
                    raise ValueError(
                        "The reference kernel is not square, and does not define a RKHS"
                    )

        self.reference_shape = (len(ref_rmean), len(ref_cmean))
        self.ref_cmean =ref_cmean
        self.ref_rmean = ref_rmean
        self.ref_mean = ref_mean
    
    return self
```

Reply:
Okay, yes, I was thinking about the same thing.